plugins/ginkgo/src/ginkgo_linear_solver.cpp

```diff
@@ -51,35 +46,108 @@ bool InternalLinearSolver::solve(const Matrix& A, const Vector& b, Vector& x)
-  /// --- Find a better way to switch solver ...
+  // create the solver
   switch (m_options.solver()) {
-  case OptionTypes::CG:
-    solve_CG(A, b, x, prec, iter_stop, res_stop, exec, time);
-    break;
-  case OptionTypes::GMRES:
-    solve_GMRES(A, b, x, prec, iter_stop, res_stop, exec, time);
-    break;
-  case OptionTypes::BICG:
-    solve_BICG(A, b, x, prec, iter_stop, res_stop, exec, time);
-    break;
-  case OptionTypes::BICGSTAB:
-    solve_BICGSTAB(A, b, x, prec, iter_stop, res_stop, exec, time);
-    break;
+  case OptionTypes::CG: {
+    // generate the solver
+    solver = gko::solver::Cg<double>::build().with_criteria(gko::share(iter_stop), gko::share(res_stop)).on(exec)->generate(pA);
+
+    // set the preconditioner
+    if (m_options.preconditioner() == OptionTypes::Jacobi) { //Jacobi
+      gko::as<gko::solver::Cg<double>>(solver)->set_preconditioner(jacobi_prec_factory->generate(pA));
+    }
+    else if (m_options.preconditioner() == OptionTypes::Ilu) { //ILU
+      gko::as<gko::solver::Cg<double>>(solver)->set_preconditioner(ilu_prec_factory->generate(gko::share(par_ilu)));
+    }
+  } break;
+  case OptionTypes::GMRES: {
+    // generate the solver
+    solver = gko::solver::Gmres<double>::build().with_criteria(gko::share(iter_stop), gko::share(res_stop)).on(exec)->generate(pA);
+
+    // set the preconditioner
+    if (m_options.preconditioner() == OptionTypes::Jacobi) { //Jacobi
+      gko::as<gko::solver::Gmres<double>>(solver)->set_preconditioner(jacobi_prec_factory->generate(pA));
+    }
+    else if (m_options.preconditioner() == OptionTypes::Ilu) { //ILU
+      gko::as<gko::solver::Gmres<double>>(solver)->set_preconditioner(ilu_prec_factory->generate(gko::share(par_ilu)));
+    }
+  } break;
+  case OptionTypes::BICG: {
+    // generate the solver
+    solver = gko::solver::Bicg<double>::build().with_criteria(gko::share(iter_stop), gko::share(res_stop)).on(exec)->generate(pA);
+
+    // set the preconditioner
+    if (m_options.preconditioner() == OptionTypes::Jacobi) { //Jacobi
+      gko::as<gko::solver::Bicg<double>>(solver)->set_preconditioner(jacobi_prec_factory->generate(pA));
+    }
+    else if (m_options.preconditioner() == OptionTypes::Ilu) { //ILU
+      gko::as<gko::solver::Bicg<double>>(solver)->set_preconditioner(ilu_prec_factory->generate(gko::share(par_ilu)));
+    }
+  } break;
+  case OptionTypes::BICGSTAB: {
+    // generate the solver
+    solver = gko::solver::Bicgstab<double>::build().with_criteria(gko::share(iter_stop), gko::share(res_stop)).on(exec)->generate(pA);
+
+    // set the preconditioner
+    if (m_options.preconditioner() == OptionTypes::Jacobi) { //Jacobi
+      gko::as<gko::solver::Bicgstab<double>>(solver)->set_preconditioner(jacobi_prec_factory->generate(pA));
+    }
+    else if (m_options.preconditioner() == OptionTypes::Ilu) { //ILU
+      gko::as<gko::solver::Bicgstab<double>>(solver)->set_preconditioner(ilu_prec_factory->generate(gko::share(par_ilu)));
+    }
+  } break;
```

Comment:
We could factorize more, by implementing a function or method that builds and sets up the solver according to its type.

Else I like the way solvers are created now.